pkg/cloud/affinity_groups_test.go

```diff
@@ -48,22 +55,79 @@ var _ = Describe("AffinityGroup Unit Tests", func() {
 		mockCtrl.Finish()
 	})
 
-	It("fetches an affinity group", func() {
-		dummies.AffinityGroup.ID = "" // Force name fetching.
-		ags.EXPECT().GetAffinityGroupByName(dummies.AffinityGroup.Name).Return(&cloudstack.AffinityGroup{}, 1, nil)
+	Context("Fetch or Create Affinity group", func() {
+		It("fetches an affinity group by Name", func() {
+			dummies.AffinityGroup.ID = "" // Force name fetching.
+			ags.EXPECT().GetAffinityGroupByName(dummies.AffinityGroup.Name).Return(&cloudstack.AffinityGroup{}, 1, nil)
+
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("fetches an affinity group by ID", func() {
+			ags.EXPECT().GetAffinityGroupByID(dummies.AffinityGroup.ID).Return(&cloudstack.AffinityGroup{}, 1, nil)
+
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("creates an affinity group", func() {
+			dummies.SetDummyDomainAndAccount()
+			dummies.SetDummyDomainID()
+			ags.EXPECT().GetAffinityGroupByID(dummies.AffinityGroup.ID).Return(nil, -1, fakeError)
+			ags.EXPECT().NewCreateAffinityGroupParams(dummies.AffinityGroup.Name, dummies.AffinityGroup.Type).
+				Return(&cloudstack.CreateAffinityGroupParams{})
+			ags.EXPECT().CreateAffinityGroup(ParamMatch(And(NameEquals(dummies.AffinityGroup.Name)))).
+				Return(&cloudstack.CreateAffinityGroupResponse{}, nil)
+
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("creates an affinity group if Name provided returns more than one affinity group", func() {
+			dummies.AffinityGroup.ID = "" // Force name fetching.
+			agp := &cloudstack.CreateAffinityGroupParams{}
+			ags.EXPECT().GetAffinityGroupByName(dummies.AffinityGroup.Name).Return(&cloudstack.AffinityGroup{}, 2, nil)
+			ags.EXPECT().NewCreateAffinityGroupParams(gomock.Any(), gomock.Any()).Return(agp)
+			ags.EXPECT().CreateAffinityGroup(agp).Return(&cloudstack.CreateAffinityGroupResponse{}, nil)
 
-		Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("creates an affinity group if getting affinity group by name fails", func() {
+			dummies.AffinityGroup.ID = "" // Force name fetching.
+			agp := &cloudstack.CreateAffinityGroupParams{}
+			ags.EXPECT().GetAffinityGroupByName(dummies.AffinityGroup.Name).Return(nil, -1, fakeError)
+			ags.EXPECT().NewCreateAffinityGroupParams(gomock.Any(), gomock.Any()).Return(agp)
+			ags.EXPECT().CreateAffinityGroup(agp).Return(&cloudstack.CreateAffinityGroupResponse{}, nil)
+
+			Ω(client.GetOrCreateAffinityGroup(dummies.AffinityGroup)).Should(Succeed())
+		})
+
+		It("creates an affinity group if ID provided returns more than one affinity group", func() {
```

Comment:
This also weirds me out.

If this test and the underlying implementation truly act this way, wouldn't this in the end return an error if working against a real ACS instance?

I would think it would then attempt to create the new Affinity group and fail due to the new name?

What do you think? If you agree this is odd, would you be up for taking on that change in implementation and the test case to make our intention here more clear?

My thinking is that this should return a nice descriptive error message that says we found more than one affinity group with that name.

Reply:
Ideally this is something that cannot happen, i.e., have 2 affinity groups with the same name - even via a DB hack, as there's a unique key constraint on the name column.